homeassistant/components/mqtt/__init__.py

```diff
@@ -783,15 +830,22 @@ def _mqtt_handle_message(self, msg) -> None:
     def _mqtt_on_disconnect(self, _mqttc, _userdata, result_code: int) -> None:
         """Disconnected callback."""
         # When disconnected because of calling disconnect()
-        if result_code == 0:
+        if result_code == MQTT_ERR_SUCCESS:
             return
 
+        self._task_reconnect = \
+            self.hass.async_create_task(self._mqtt_reconnect(result_code))
+
+    async def _mqtt_reconnect(self, result_code) -> None:
+        """Reconnect to broker."""
         tries = 0
 
         while True:
             try:
-                if self._mqttc.reconnect() == 0:
+                if await self.hass.async_add_job(self._mqttc.reconnect) == \
```

Comment:
If it does blocking I/O we should schedule it on the executor thread pool via `hass.async_add_executor_job`.

Reply:
I did some more digging, the SSL handshake should not block: https://github.com/eclipse/paho.mqtt.python/pull/124

Reply:
Ok. Is there any I/O here? If so we should still schedule it on the executor. But if all the I/O is handled deeper down by the socket and select/loop, we could perhaps just treat it as a callback and call it directly?

Reply:
The potentially blocking I/O is the SSL handshake, but I'm not sure if it blocks or not @_@

Reply:
We call the client `connect` method directly, right? And that is calling `reconnect` internally.

Reply:
Good point! Then it's fine with the `hass.async_add_job()`?

Reply:
https://pypi.org/project/paho-mqtt/#connect-reconnect-disconnect

> The connect() function connects the client to a broker. This is a blocking function.

Then I'm assuming `reconnect` is too. We could use `connect_async` instead.

Reply:
`hass.async_add_job` is legacy. We should use `hass.async_add_executor_job` if we know that we need the executor.

Reply:
I suggest we use `connect_async` instead of `connect`, and call this directly, and schedule `reconnect` on the executor.

Reply:
The difference between `connect()` and `connect_async()` is in fact only that the former calls `reconnect()`, while the latter does not.

So to conclude, I should update to:
- Call `connect_async()` instead
- Call `reconnect()` through `hass.async_add_executor_job`

Right?

Reply:
Yes!

Reply:
Updated such that reconnect() is called through hass.async_add_executor_job

However, instead of using connect_async() I call connect() through hass.async_add_executor_job, because this is the implementation of connect():
```python
def connect(self, host, port=1883, keepalive=60, bind_address=""):
        self.connect_async(host, port, keepalive, bind_address)
        return self.reconnect()
```

Otherwise, we could do like this to initiate the reconnect():
```python
self._mqttc.connect_async()
self._mqtt_on_disconnect(mqtt.MQTT_ERR_SUCCESS)
```